### PR TITLE
Automate per-package release for specific components

### PR DESCRIPTION
**Pull request by lmolkova**
# Description

Adds automation allowing to release individual components.
Packages that can be released individually are:
- excluded from the main release
- must have individual changelog
- follow the same release process and versioning rules:
  - prepare minor/patch that creates 
    - long-term `package-release/{package}/v{version}` branch
    - PRs to update changelog and versions in release branch and main
  - release
- adds changelogs and versions for for components that are currently released manually as individual packages (`opentelemetry-resource-detector-azure`, `opentelemetry-sdk-extension-aws`, `opentelemetry-propagator-aws-xray`)

Related to #2858

## Type of change

# How Has This Been Tested?

Tested in the fork https://github.com/lmolkova/opentelemetry-python-contrib/actions

- [x] Preparing + releasing new minor or patch version of an unstable package 
- [x] Preparing + releasing 'common' 

# Does This PR Require a Core Repo Change?

- [x] No.

# Checklist:

See [contributing.md](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/CONTRIBUTING.md) for styleguide, changelog guidelines, and more.

- [x] Followed the style guidelines of this project
- [ ] Changelogs have been updated

### CODE REVIEW

File: .github/workflows/package-prepare-patch-release.yml
Comment (lmolkova) [inline .github/workflows/package-prepare-patch-release.yml:9]: TODO: add `opentelemetry-propagator-aws-xray`

File: .github/workflows/package-prepare-release.yml
Comment (lmolkova) [inline .github/workflows/package-prepare-release.yml:9]: TODO: add `opentelemetry-propagator-aws-xray`

File: .github/workflows/package-release.yml
Comment (lmolkova) [inline .github/workflows/package-release.yml:9]: TODO: add `opentelemetry-propagator-aws-xray`

File: CHANGELOG.md
Comment (lzchen) [inline CHANGELOG.md:8]: Do you envision this to be a temporary place to hold this note? Will we constantly be updating this when gen-ai instrumentations are added? Perhaps we put all gen-ai instrumentations in the future in a specific folder and outline that components in that folder will also have individual changelogs?

File: RELEASING.md
Comment (lzchen) [inline RELEASING.md:18]: Maybe we can add a section in RELEASING.md outlining which packages are supported and link to there from this note?

File: CHANGELOG.md
Comment (lmolkova) [reply CHANGELOG.md:8]: I think that in the long term we should support individual release for all components (including existing non-gen-ai ones). 

I.e. the central CHANGELOG would disappear. 

I'm going to steal the following approach from [js-contrib](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/ee5c584c943eb8a3de0566843a42784f16c0f31f/CHANGELOG.md?plain=1#L4):

> ~~As of v0.25.1 (2022-01-24) changelog content has moved to separate CHANGELOG.md files for each package.~~ Use [this search for a list of all CHANGELOG.md files in this repo](https://github.com/search?q=repo%3Aopen-telemetry%2Fopentelemetry-js-contrib+path%3A**%2FCHANGELOG.md&type=code).

File: RELEASING.md
Comment (lzchen) [inline RELEASING.md:76]:
```suggestion
Per-package release is handled by the [`[Package] Release`](./.github/workflows/package-release.yml) workflow that allows
```

File: .github/workflows/package-prepare-release.yml
Comment (lzchen) [inline .github/workflows/package-prepare-release.yml:54]: Nit: Should we verify that there is a CHANGELOG as well?

File: .github/workflows/package-prepare-release.yml
Comment (lzchen) [inline .github/workflows/package-prepare-release.yml:109]: Nit: Can we use Python 3.11?

File: scripts/generate_release_notes.sh
Comment (lzchen) [inline scripts/generate_release_notes.sh:6]: Nit: Can we include that this script is only being used for package-release.yml workflow?

File: scripts/merge_changelog_to_main.sh
Comment (lzchen) [inline scripts/merge_changelog_to_main.sh:4]: Nit: Doesn't this also update the actual CHANGELOG as well?

**Review by lzchen — APPROVED** [review]
LGTM! Thanks for working on this! Just a few minor comments. Feel free to reach out to one of the maintainers to actually test this out with an actual package once it is merged.

File: .github/workflows/package-prepare-release.yml
Comment (lmolkova) [reply .github/workflows/package-prepare-release.yml:51]: Let's merge https://github.com/open-telemetry/opentelemetry-python/pull/4099 since it seems to be very close and I'll update this one

File: scripts/generate_release_notes.sh
Comment (lmolkova) [reply scripts/generate_release_notes.sh:6]: good catch,  I was going to reuse this script in general release script too - updated the PR and tested it in the fork.

File: .github/workflows/package-prepare-release.yml
Comment (lmolkova) [reply .github/workflows/package-prepare-release.yml:59]: Added on line ~37

File: .github/workflows/package-prepare-release.yml
Comment (lmolkova) [reply .github/workflows/package-prepare-release.yml]: I just realized that https://github.com/open-telemetry/opentelemetry-python/pull/4099 is in the core repo and there is no PR against contrib.

I updated workflow to support both - `version.py` and `version/__init__py` so it will keep working when contrib is changed to `version/__init__py`. It should be easy to become strict again later one. 

So I'm going to close this one.

File: .github/workflows/release.yml
Comment (lmolkova) [inline .github/workflows/release.yml:1]: need to test if general prepare-release workflow does not update dependency versions for excluded packages - see 
https://github.com/open-telemetry/opentelemetry-python-contrib/pull/2759#discussion_r1777421590 for the context

Comment (lmolkova) [reply .github/workflows/release.yml:1]: tried in https://github.com/lmolkova/opentelemetry-python-contrib/actions/runs/11057404852, versions are not updated for excluded packages.

**Comment by lmolkova** [comment]
@open-telemetry/opentelemetry-python-contrib-approvers could I get some eyes on it please?  Thank you!

File: propagator/opentelemetry-propagator-aws-xray/src/opentelemetry/propagators/aws/version.py
Comment (xrmx) [inline propagator/opentelemetry-propagator-aws-xray/src/opentelemetry/propagators/aws/version.py:15]: Has this been bumped to 1.1.0.dev because these changes are thought as adding features right?

File: .github/workflows/package-prepare-patch-release.yml
Comment (aabmass) [inline .github/workflows/package-prepare-patch-release.yml:1]: Do we need a separate workflow for minor releases?

**Review by aabmass — COMMENTED** [review]
Thanks for working on it @lmolkova

It seems reasonable to me and I think @lzchen has the most context and has already approved. I'm OK to go ahead with this.

File: propagator/opentelemetry-propagator-aws-xray/src/opentelemetry/propagators/aws/version.py
Comment (lmolkova) [reply propagator/opentelemetry-propagator-aws-xray/src/opentelemetry/propagators/aws/version.py:15]: Since we had no automation for these packages in the past, we currently have the version that's already been released -https://pypi.org/project/opentelemetry-propagator-aws-xray/1.0.2/

For packages that are released with workflow we use the vNext (minor + 1). 

So I'm just mimicking the same state/strategy as if it was released automatically:

- next version is assumed to be minor+1 (expecting new features and bot expecting patches)
- patch release should be done from the long-lived release branch. It should be possible to backport fixes to that branch when necessary.

Having said that, it's possible to release a patch from the main branch by updating the version on main to `1.0.3.dev` and running the release workflow - it'll prepare release `1.0.3` and send PR to bump version to `1.1.0.dev` - here's an example https://github.com/lmolkova/opentelemetry-python-contrib/actions/runs/11225953867.

File: .github/workflows/package-prepare-patch-release.yml
Comment (lmolkova) [reply .github/workflows/package-prepare-patch-release.yml:1]: I just replicated existing release process that we have for 'genera' release to individual components, the existing workflows have different behavior:
- **prepare release** runs from main, prepares release branch and also sends PR against main to bump the next version to minor + 1
- **prepare patch** runs from the **release** branch. E.g.:
   -  I released 1.2.0 - it has  long-lived `release/1.2.x` branch. 
   -  Now I'm working on 1.3.0 and discovered a bug
   - I'd patch it on main
   - And can decide to backport it to `release\1.2.x`
   - Once backported fix is merged I can release 1.2.1 from `release/1.2.x` branch. 
   - I don't need to bump component version on main or update changelog there

As a result, these two workflows are somewhat different. 
This is documented in https://github.com/open-telemetry/opentelemetry-python-contrib/blob/main/RELEASING.md and also it a common pattern across otel documented here https://github.com/trask/repository-practices

File: propagator/opentelemetry-propagator-aws-xray/src/opentelemetry/propagators/aws/version.py
Comment (xrmx) [reply propagator/opentelemetry-propagator-aws-xray/src/opentelemetry/propagators/aws/version.py:15]: Thanks for the explanation

**Review by xrmx — APPROVED** [review]

**Pull request merged**
